core/src/test/java/org/infinispan/scattered/statetransfer/CoordStopTest.java

```diff
@@ -0,0 +1,166 @@
+import org.infinispan.topology.CacheTopology;
+import org.infinispan.util.BlockingLocalTopologyManager;
+import org.infinispan.util.BlockingLocalTopologyManager.BlockedTopology;
+import org.infinispan.util.ControlledConsistentHashFactory;
+import org.infinispan.util.ControlledRpcManager;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.Test;
+
+@Test(groups = "functional", testName = "scattered.statetransfer.CoordStopTest")
+@CleanupAfterMethod
+public class CoordStopTest extends MultipleCacheManagersTest {
+   private CountDownLatch viewLatch = new CountDownLatch(1);
+
+   @Override
+   public Object[] factory() {
+      return new Object[] {
+            new CoordStopTest().biasAcquisition(BiasAcquisition.NEVER),
+            new CoordStopTest().biasAcquisition(BiasAcquisition.ON_WRITE)
+      };
+   }
+
+   @Override
+   protected void createCacheManagers() throws Throwable {
+      ConfigurationBuilder cb = new ConfigurationBuilder();
+      cb.clustering().cacheMode(CacheMode.SCATTERED_SYNC).hash().numSegments(3);
+      if (biasAcquisition != null) {
+         cb.clustering().biasAcquisition(biasAcquisition);
+      }
+      ControlledConsistentHashFactory.Scattered chf = new ControlledConsistentHashFactory.Scattered(new int[] {0, 1, 2});
+      cb.clustering().hash().consistentHashFactory(chf);
+
+      addClusterEnabledCacheManager(cb);
+      addClusterEnabledCacheManager(cb);
+      // If the updated topologies from old coord come when it's no longer in the view these are ignored.
+      // Therefore we have to delay the view.
+      GlobalConfigurationBuilder gcb = new GlobalConfigurationBuilder();
+      gcb.transport().transport(new DelayedViewJGroupsTransport(viewLatch));
+      addClusterEnabledCacheManager(gcb, cb);
+
+      assertTrue(cache(0).getCacheManager().isCoordinator());
+      // start other caches
+      cache(1);
+      cache(2);
+      waitForClusterToForm();
+
+      // segment 0 will be moved to cache(2). Since we've lost coord cache(1) now -> 0 and cache(2) -> 1
+      chf.setOwnerIndexes(new int[][] { {1}, {0}, {1} });
+   }
+
+   @AfterMethod
+   @Override
+   protected void clearContent() throws Throwable {
+      super.clearContent();
+      viewLatch = new CountDownLatch(1);
+   }
+
+   // Reproducer for ISPN-9128
+   public void testCoordLeaves() throws InterruptedException, ExecutionException {
+      MagicKey key = new MagicKey(cache(0));
+      cache(1).put(key, "value");
+      int stableTopologyId = cache(1).getAdvancedCache().getDistributionManager().getCacheTopology().getTopologyId();
+      BlockingLocalTopologyManager ltm2 = BlockingLocalTopologyManager.replaceTopologyManager(manager(2), cache(2).getName());
+      ControlledRpcManager rpcManager2 = ControlledRpcManager.replaceRpcManager(cache(2));
+      rpcManager2.excludeCommands(cmd -> cmd instanceof StateResponseCommand, CoordStopTest::isCancelStateTransfer);
```

Comment:
IMO it would be simpler to handle the response commands explicitly, then you can guarantee there are no cancel commands.

Reply:
What do you mean by handling explicitly? Overriding RpcManager on my own?

Reply:
No, calling `ControlledRpcManager.expectCommand()`

Reply:
IIUC the ControlledRpcManager requires the commands to be send in order, and cancel commands depend on the timing - I would probably need to add additional synchronization to enforce them. Not that it wouldn't be possible, but I simply don't care about them.

Reply:
IMO the order of the cancel commands is pretty straightforward: if you intercept the response commands and you wait for their responses, all the segments will be received and there will be no need for a cancel command.